<launch>
  <arg name="imu_enabled" default="true" />
  <arg name="imu_rate" default="5.0" />
  <arg name="battery_enabled" default="true" />
  <arg name="battery_rate" default="0.5" />
  <arg name="lights_enabled" default="true" />
  <arg name="lights_rate" default="0.5" />
  <arg name="nav_enabled" default="true" />
  <arg name="nav_rate" default="5.0" />

  <arg name="namespace" default="$(optenv ROBOT_NAMESPACE)" />

  <group ns="/$(arg namespace)">
    <group ns="nmea_if">

      <!-- Relay sentences from GPS receiver -->
      <node pkg="topic_tools" type="relay" name="navsat_sentence_relay"
            args="/$(arg namespace)/navsat/nmea_sentence tx" />

      <!-- GPS time offset calculator -->
      <param name="time_offset_default" value="-14401.595" />
      <node pkg="heron_nmea" type="gps_time_offset_publisher" name="time_offset">
        <remap from="rx" to="navsat/rx" />
      </node>

      <!-- Publishes $CPNVG and $CPNVR messages -->
      <node pkg="topic_tools" type="throttle" name="odom_throttle"
            args="messages /$(arg namespace)/odometry/filtered $(arg nav_rate) odom"
            if="$(arg nav_enabled)" />
      <node pkg="heron_nmea" type="nav_publisher" name="nmea_nav">
        <param name="namespace" value="$(arg namespace)" />
        <remap from="nmea_if/navsat/vel" to="navsat/vel" />
        <remap from="nmea_if/navsat/fix" to="navsat/fix" />
      </node>

      <!-- Publishes $CPIMU and $CPRCM messages -->
      <node pkg="topic_tools" type="throttle" name="imu_throttle"
            args="messages /$(arg namespace)/imu/data $(arg imu_rate) imu/data"
            if="$(arg imu_enabled)" />
      <node pkg="heron_nmea" type="imu_publisher" name="nmea_imu">
        <param name="namespace" value="$(arg namespace)" />
      </node>

      <!-- Publishes $CPRBS messages -->
      <node pkg="topic_tools" type="throttle" name="battery_throttle"
            args="messages /$(arg namespace)/sense $(arg battery_rate) sense"
            if="$(arg battery_enabled)" />
      <node pkg="heron_nmea" type="battery_publisher" name="nmea_battery" />

      <!-- Publishes $CPALT messages -->
      <node pkg="topic_tools" type="throttle" name="lights_throttle"
            args="messages /$(arg namespace)/lights $(arg lights_rate) lights"
            if="$(arg lights_enabled)" />
      <node pkg="heron_nmea" type="lights_publisher" name="nmea_lights" />

      <!-- Receives $PYDIR, $PYDEV, and $PYDEP messages -->
      <node pkg="heron_nmea" type="command_publisher" name="nmea_commands">
        <remap from="cmd_drive" to="/$(arg namespace)/cmd_drive" />
        <remap from="cmd_helm" to="/$(arg namespace)/cmd_helm" />
        <remap from="cmd_course" to="/$(arg namespace)/cmd_course" />
        <remap from="disable_lights" to="/$(arg namespace)/disable_lights" />
        <param name="namespace" value="$(arg namespace)" />
      </node>

      <!-- NMEA HERON interface -->
      <node pkg="nmea_comms" type="socket_node" name="socket_node">
        <remap from="nmea_sentence_out" to="tx" />
        <remap from="nmea_sentence" to="rx" />
      </node>
    </group>
  </group>
</launch>
